errors/InvalidInstanceHead.md

```diff
@@ -2,20 +2,129 @@
 
 ## Example
 
+As of 0.11.6, the `InvalidInstanceHead` error has been added to the
+purescript compiler. The error can be summoned using the following:
+
 ```purescript
-module ShortFailingExample where
+module InvalidInstanceHead where
+
+class ClassWithInvalidInstanceHead (x :: # Type)
 
+instance FailingExample :: ClassWithInvalidInstanceHead ()
 ...
 ```
 
 ## Cause
 
-Explain why a user might see this error.
+Normally, when this error rears its head, it's because you're trying to write
+an `instance` of a given `class` that pattern matches on some **row type**. The
+issue here is that there isn't really a sensible way to match on records. There
+are, however, ways to get around this limitation.
 
 ## Fix
 
-- Suggest possible solutions.
+With the arrival of [`purescript-typelevel-prelude`](https://github.com/purescript/purescript-typelevel-prelude)
+2.3.0 came the introduction of `RowToList` and its surrounding classes. The
+basic idea is that you can convert a row type to a type-level cons list of its
+entries. For example:
+
+```purescript
+--        ROW              LIST
+
+RowToList ()               Nil
+RowToList (a :: A)         (Cons "a" A Nil)
+RowToList (a :: A, b :: B) (Cons "a" A (Cons "b" B Nil))
+```
+
+The `RowToList` constraint allows us to derive, from a row type, the `RowList`
+that complements it. The keys of the `RowList` are alphabetically ordered, which
+allows us to manipulate row types in a deterministic way. If your instance heads
+pattern match on `RowList` instead of `# Type`, this error goes away.
+
+## Worked Example
+
+Let's imagine a class that we use to assert that a record is homogenous. Or, in
+simpler terms, all the fields in the record have the same type.
+
+```purescript
+class IsHomogenousRowType (rowType :: # Type)
+```
+
+Now, as we've seen earlier, we can't just pattern-match on this in its current
+state; we need to convert it to a `RowList`. How do we do that?
+
+```purescript
+instance isHomogenousImpl
+  :: ( RowToList rowType rowList
+     , IsHomogenousRow rowList a
+     )
+  => IsHomogenousRecord rowType
+```
+
+Here, we say that, "if you can find me a `RowList` representation for my record
+that satisfies `IsHomogenousRow`, then this record is good to go". Notice how
+we're "pattern matching" on the left side of `RowToList` - we don't know what
+`la` is, but we trust the compiler to iron out that little detail for us.
+
+Now that's all ready to roll, we should probably write `IsHomogenousRow`:
+
+```purescript
+class IsHomogenousRow (list :: RowList) a | list -> a
+```
+
+Because we're now dealing with a `RowList`, we can pattern match and avoid those
+pesky errors about instance heads. Our functional dependency here says that, "if
```

Comment:
I think this explanation of the functional dependency can be deleted - it isn't any more informative that the definition of a functional dependency, which can be found elsewhere in this documentation or wikipedia.

Reply:
I don't mind the explanation here. I've definitely found myself chasing definitions until I didn't know where I started anymore before.